Assets/Scripts/Managers/EnemySpawner.cs

```diff
@@ -183,4 +140,101 @@ private ChessPiece SpawnEnemy(ChessPieceType pieceType, IndexCode indexCode)
     }
+
+    private ChessPieceType GetRandomSpawningPiece()
+    {
+        // Get an array of all the ChessPiece types
+        Array pieceValues = Enum.GetValues(typeof(ChessPieceType));
+        ChessPieceType pieceType = (ChessPieceType)pieceValues.GetValue(random.Next(pieceValues.Length));
+
+        // If piece type is king, iterate until other ChessPieceType is found
+        while (pieceType == ChessPieceType.King)
+        {
+            pieceType = (ChessPieceType)pieceValues.GetValue(random.Next(pieceValues.Length));
+        }
+
+        return pieceType;
+    }
+
+    private bool SpawnRandomlyOnBoard(bool enableUpgrades)
+    {
+        int turnNumber = _gameStateManager.Value.GetTurnNumber();
+        ChessPieceType randomPiece = GetRandomSpawningPiece();
+        int rowForSpawning = random.Next(5, 7);
+        int columnForSpawning = random.Next(0, _boardManager.Value.GetBoardDimensions()[0]);
+        IndexCode position = _boardRows[rowForSpawning][columnForSpawning];
+                
+        // Create chess piece. If the board square is already occupied, loop until a valid board square is found.
+        ChessPiece piece = null;
+        int counter = 0;
+        while (piece == null)
+        {
+            counter++;
+            piece = SpawnEnemy(randomPiece, position);
+
+            if (piece)
+            {
+                Debug.Log($"Spawning enemy chess piece of type {randomPiece}, on IndexCode {position}");
+                // Avoid ChessPïece functions division by zero
+                piece.Range = 1;
+                piece.Speed = 1;
+                break;
+            }
+                    
+            // Relocate chess piece position if it is not valid
+            rowForSpawning = random.Next(5, 7);
+            columnForSpawning = random.Next(0, _boardManager.Value.GetBoardDimensions()[0]);
+            position = _boardRows[rowForSpawning][columnForSpawning];
+                    
+            // If all square boards are occupied, no spawning is done
+            // This prevents an infinite loop if all board squares have their pieces assigned
+            if (counter > 24)
+            {
+                // If no squares are available, do not spawn
+                Debug.Log("Enemy could not be created because there is not a valid board square.");
+                return false;
+            }
+        }
+        
+        if (piece && enableUpgrades)
+        {
+            Debug.Log("Applying upgrades to spawned piece.");
+            // Random upgrades
+            double rangeUpgradeChance = random.NextDouble();
+            double speedUpgradeChance = random.NextDouble();
+            // Only upgrade pieces after turn 3
+            if (turnNumber < 3)
+            {
+                // do nothing
+            }
+            else if (turnNumber < 10)
+            {
+                if (rangeUpgradeChance < 0.3)
+                    piece.Range++;
+                if (speedUpgradeChance < 0.5)
```

Comment:
It seems like upgrades can only happen once per piece?

So at best they will have +1 range and +1 speed?

I think would be good to allow for multiple upgrades as well. e.g. +3 range and +3 speed etc.

Reply:
Upgrades can now be at maximum +4 +4 after turn 20.